File: web/src/lib/realtime.ts
"use client";

/**
 * The lobby, streamed.
 *
 * This is the frontend's only connection to Supabase: a WebSocket carrying one
 * message per change, on a channel named after the lobby.
 *
 * It used to be a doorbell — the payload was `{"version": 7}` and every client
 * answered it with an HTTP fetch. That made the socket a *trigger* for a poll
 * rather than a replacement for one, and left the round trip it existed to
 * remove exactly where it was. Now the message carries the state and the room
 * renders it as it lands.
 *
 * What arrives is `LobbyView`, the same redacted object `GET /lobbies/:code`
 * returns — answers still in the pool do not name their category and no
 * explanation appears before its round is over. The API decides that, here as
 * before; this file only unwraps what it sent.
 *
 * It is still *not* a database connection. Nothing here reads a table, and
 * nothing could: the anon key it holds is subject to row level security with no
 * policies, so it returns zero rows from every table in the schema.
 *
 * Deliberately `@supabase/realtime-js` rather than `@supabase/supabase-js`.
 * The full SDK bundles clients for Postgres, auth, storage and edge functions
 * -- roughly 4.5 MB of dependency for features this app never calls, some of
 * which `createClient` instantiates whether you use them or not. This is the
 * one piece that does the job, and depending on only it keeps the frontend
 * honest about what it talks to.
 *
 * Everything here is best effort. If Supabase is not configured, or the socket
 * never connects, `subscribe` returns a no-op teardown and the room keeps
 * working on its slower poll. A realtime outage should cost latency, not a
 * game.
 */

import { RealtimeClient, type RealtimeChannel } from "@supabase/realtime-js";

import type { LobbyView } from "@/lib/types";

const url = process.env.NEXT_PUBLIC_SUPABASE_URL;
const anonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

export const realtimeConfigured = Boolean(url && anonKey);

let client: RealtimeClient | null = null;

function getClient(): RealtimeClient | null {
  if (!realtimeConfigured) return null;
  client ??= new RealtimeClient(`${url!.replace(/\/+$/, "")}/realtime/v1`, {
    params: { apikey: anonKey! },
  });
  return client;
}

/**
 * Call `onChange` whenever the lobby changes, with the new state when it came
 * along. Returns a teardown function.
 *
 * `view` is null when the API sent the version alone, which it does for a state
 * too large to put in a frame -- a long game accumulates an answer key per
 * finished round. The caller fetches in that case, exactly as it used to for
 * every message.
 *
 * Safe to call when realtime is not configured -- it simply never fires.
 */
export function subscribeToLobby(
  code: string,
  onChange: (view: LobbyView | null) => void,
): () => void {
  const socket = getClient();
  if (!socket) return () => {};

  const channel: RealtimeChannel = socket.channel(`lobby:${code.toUpperCase()}`);
  channel
    .on("broadcast", { event: "lobby-changed" }, (message) => {
      const state = (message?.payload as { state?: unknown } | undefined)?.state;
      const view =
        state && typeof state === "object" && "version" in state
          ? (state as LobbyView)
          : null;
      onChange(view);
    })
    .subscribe();

  return () => {
    void socket.removeChannel(channel);
  };
}

/**
 * Picks, as they are made, on everybody's screen.
 *
 * A player considering an answer is not a move -- the move is the placement,
 * and that still goes over HTTP and comes back as a `LobbyView` like every
 * other one. But the table is watching one player think, and until now it had
 * nothing to watch: the chip they clicked lit up in their own browser and
 * nowhere else. This carries that click, and only that click.
 *
 * Its own topic rather than `lobby:CODE`, which belongs to the API -- clients
 * only ever listen there. Two topics also means the server is never taught an
 * event it has no opinion about, and joining one topic twice on a single socket
 * is a duplicate join.
 *
 * Nothing here is authoritative. The payload is an item id the receiver already
 * has in its pool, sent by someone who already holds the lobby code; a forged
 * one flashes a chip and changes no state. A dropped frame costs a flash.
 */
const PICK_EVENT = "answer-picked";

export function subscribeToPicks(
  code: string,
  onPick: (itemId: string) => void,
): { send: (itemId: string) => void; close: () => void } {
  const socket = getClient();
  if (!socket) return { send: () => {}, close: () => {} };

  /* `self: false` is the default, and is set anyway: the picker's own chip has
     already flared off the click, without waiting for a round trip. */
  const channel: RealtimeChannel = socket.channel(`lobby:${code.toUpperCase()}:picks`, {
    config: { broadcast: { self: false } },
  });
  channel
    .on("broadcast", { event: PICK_EVENT }, (message) => {
      const itemId = (message?.payload as { item_id?: unknown } | undefined)?.item_id;
      if (typeof itemId === "string") onPick(itemId);
    })
    .subscribe();

  return {
    send: (itemId) =>
      void channel.send({
        type: "broadcast",
        event: PICK_EVENT,
        payload: { item_id: itemId },
      }),
    close: () => void socket.removeChannel(channel),
  };
}

File: api/app/services/realtime.py
"""Telling players that something changed, without them having to ask.

The game used to be discovered by polling: every client asked for the whole
lobby every 1.5 seconds whether anything had happened or not. That is a
read-modify-write against the shared store per client per interval -- tens of
thousands an hour for a four-player game, nearly all of them answering
"nothing new".

So the API says when something happened instead. After every mutation it
publishes one tiny message on a Supabase Realtime channel named after the lobby,
and clients fetch only when they hear one.

**The message carries the state.** This was a doorbell to begin with -- payload
`{"version": 7}`, and every client answered it with a fetch. That made a socket
the trigger for a poll rather than a replacement for one: the round trip it
existed to remove was still there, just moved. So the new state goes in the
message and clients render it as it arrives.

What goes out is `LobbyView`, not `Lobby`. The distinction is the whole safety
argument: `Lobby` holds `ItemSolution.category_id` and `explanation` for every
answer -- the complete solution -- while `LobbyView` is what the API already
hands anyone who asks for the lobby, with unplaced answers redacted. It is also
identical for every player: `_view` takes no player id, so there is no per-player
redaction to get wrong here.

That settles who can listen. The channel is named after the lobby code, and
anyone with the code can already call the API and get exactly this. Listening
tells them nothing extra.

Oversized views fall back to the doorbell. `finished_rounds` accumulates a full
answer key per round, so a long game's view can grow past what a broadcast
frame should carry; past `MAX_PAYLOAD_BYTES` the version goes out alone and the
client fetches, as it used to.

Publishing is best-effort by construction. A move that succeeded must not fail
because a notification did not go out -- clients keep a slow poll as a backstop,
so a dropped message costs latency and nothing else.
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING

import httpx

from app.config import get_settings

if TYPE_CHECKING:
    from app.schemas import LobbyView

logger = logging.getLogger(__name__)

TIMEOUT_SECONDS = 2.0

EVENT = "lobby-changed"

MAX_PAYLOAD_BYTES = 64 * 1024


def channel_for(code: str) -> str:
    return f"lobby:{code.upper()}"


def publish(view: LobbyView) -> bool:
    """Send `view` to everyone watching its lobby. Never raises.

    Returns whether the message went out, which is used by tests and by nothing
    else -- no caller should change behaviour based on it.
    """
    settings = get_settings()
    if not settings.realtime_broadcast:
        return False
    if not settings.supabase_url or not settings.supabase_service_role_key:
        return False

    url = f"{settings.supabase_url.rstrip('/')}/realtime/v1/api/broadcast"
    key = settings.supabase_service_role_key

    payload: dict[str, object] = {"version": view.version}
    state = view.model_dump(mode="json")
    if len(json.dumps(state)) <= MAX_PAYLOAD_BYTES:
        payload["state"] = state

    try:
        response = httpx.post(
            url,
            headers={
                "apikey": key,
                "Authorization": f"Bearer {key}",
                "Content-Type": "application/json",
            },
            json={
                "messages": [
                    {
                        "topic": channel_for(view.code),
                        "event": EVENT,
                        "payload": payload,
                    }
                ]
            },
            timeout=TIMEOUT_SECONDS,
        )
        response.raise_for_status()
        return True
    except Exception as exc:  # noqa: BLE001 - a missed notification is not a failed move
        logger.debug("realtime broadcast for %s failed: %s", view.code, exc)
        return False
